File: special_shop.java

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.math.BigInteger;
import java.math.MathContext;
import java.math.BigDecimal;

class TestClass {

    private static class Parabola {
        public Parabola(BigInteger numPots, BigInteger xMult, BigInteger yMult) {
            coeffA = xMult.add(yMult);
            coeffB = yMult.multiply(numPots).multiply(BigInteger.valueOf(-2));
            coeffC = numPots.pow(2).multiply(yMult);
        }

        public BigDecimal calcMin() {
            final BigInteger numerator = coeffB.multiply(BigInteger.valueOf(-1));
            final BigInteger denominator = coeffA.multiply(BigInteger.valueOf(2));
            final BigDecimal realNumerator = new BigDecimal(numerator);
            final BigDecimal realDenominator = new BigDecimal(denominator);
            final BigDecimal realResult = 
                realNumerator.divide(realDenominator, roundingMode.getRoundingMode());
            return realResult.round(roundingMode);
        }

        public BigDecimal calcAt(BigDecimal xCoord) {
            final BigDecimal aTerm = xCoord.pow(2).multiply(new BigDecimal(coeffA));
            final BigDecimal bTerm = xCoord.multiply(new BigDecimal(coeffB));
            final BigDecimal cTerm = new BigDecimal(coeffC);
            return aTerm.add(bTerm).add(cTerm).round(roundingMode);
        }

        private BigInteger coeffA;
        private BigInteger coeffB;
        private BigInteger coeffC;

        private MathContext roundingMode = new MathContext(0);
    }


    private static BigInteger solveCase(BigInteger numPots, BigInteger xMult, BigInteger yMult) {
        final Parabola eq = new Parabola(numPots, xMult, yMult);
        final BigDecimal optimalXCoord = eq.calcMin();
        final BigDecimal result = eq.calcAt(optimalXCoord);
        return result.toBigInteger();
    }

    private static void parseTestCase(String testCaseLine, BufferedWriter out) throws IOException {
        final String[] testCase = testCaseLine.trim().split(" ");
        final BigInteger pots_to_buy = new BigInteger(testCase[0]);
        final BigInteger x_cost = new BigInteger(testCase[1]);
        final BigInteger y_cost = new BigInteger(testCase[2]);
        final BigInteger result = solveCase(pots_to_buy, x_cost, y_cost);
        out.write(result.toString());
        out.newLine();
    }

    public static void main(String args[] ) throws Exception {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final String numLinesLine = br.readLine().trim();
        final int numLines = Integer.parseInt(numLinesLine);

        final BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));
        for(int i = 0; i < numLines; ++i) {
            parseTestCase(br.readLine(), out);
        }

        out.flush();
    }
}

File: special_shop.kt

import kotlin.math.round;

class Parabola(potCount: Long, xMult: Long, yMult: Long) {
    private val coeffA = xMult + yMult;
    private val coeffB = -2 * yMult * potCount;
    private val coeffC = potCount * potCount * yMult;

    fun minXCoord(): Long {
        val numerator = -1.0 * coeffB;
        val denominator = coeffA * 2.0;
        val result = round(numerator / denominator);
        return result.toLong();
    }

    fun solveAt(xCoord: Double): Long {
        val termA = xCoord * xCoord * coeffA;
        val termB = xCoord * coeffB;
        val termC = coeffC.toDouble();
        val realResult = termA + termB + termC;
        return realResult.toLong();
    }
}


fun solveCase(potCount: Long, xMult: Long, yMult: Long): Long {
    val eq = Parabola(potCount, xMult, yMult);
    val optimal_a_pots: Long = eq.minXCoord();
    val soln: Long = eq.solveAt(optimal_a_pots.toDouble());
    return soln;
}

fun parseTestCase(): Unit {
    val caseParams = readLine()!!.split(" ");
    val potsToBuy = caseParams[0].toLong();
    val xCost = caseParams[1].toLong();
    val yCost = caseParams[2].toLong();
    val result = solveCase(potsToBuy, xCost, yCost);
    println(result);
}

fun main() {
    val caseCount = readLine()!!.toInt();

    for (i in 1..caseCount) {
        parseTestCase();
    }
}